sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SQLHadoopMapReduceCommitProtocol.scala

```diff
@@ -55,7 +55,8 @@ class SQLHadoopMapReduceCommitProtocol(
         // The specified output committer is a FileOutputCommitter.
         // So, we will use the FileOutputCommitter-specified constructor.
         val ctor = clazz.getDeclaredConstructor(classOf[Path], classOf[TaskAttemptContext])
-        committer = ctor.newInstance(new Path(path), context)
+        val committerOutputPath = if (dynamicPartitionOverwrite) stagingDir else new Path(path)
+        committer = ctor.newInstance(committerOutputPath, context)
```

Comment:
Is it the same if we pass the `committerOutputPath` in `InsertIntoHadoopFsRelationCommand` to `SQLHadoopMapReduceCommitProtocol` directly via `FileCommitProtocol.instantiate()`?

Reply:
OK

Reply:
In my opinion, it is not the same. Since for  ` SQLHadoopMapReduceCommitProtocol` which derived from `HadoopMapReduceCommitProtocol`, final output path for `dynamicPartitionOverwrite` job is the same as other kind of job, there are 2 steps during job committing:
1. move task committed directories to `/path/to/output/.spark-staging-{jobId}`, the code here inits a committer with output path `/path/to/output/.spark-staging-{jobId}`
2. move partition directories under `/path/to/output/.spark-staging-{jobId}` to final output path

Reply:
I mean, previously, we pass `outputPath` to `FileCommitProtocol.instantiate` and `qualifiedOutputPath` to `OutputSpec`. And now, you pass the qualified `committerOutputPath` (the one in `InsertIntoHadoopFsRelationCommand`) to `OutputSpec`, but you still pass `outputPath` to `FileCommitProtocol.instantiate`. So can we just pass the non-qualified `committerOutputPath` to `FileCommitProtocol.instantiate` or event the qualified one if it's the same?

Reply:
It's not the same. It's a bit confused. I think there are 2 kinds of 'committer':
1. `HadoopMapReduceCommitProtocol`, this is implemented by Spark, `outputPath` of this committer should be the `finalPath`, like `/path/to/output/`. In this patch, `FileCommitProtocol.instantiate` creates a instance of `HadoopMapReduceCommitProtocol`
2. `FileOutputCommitter`, this is implemented by hadoop mapreduce client, `outputPath` of this committer should be `intermediatePath`, in this patch, it's like `/path/to/output/.spark-staging-{jobId}`. `HadoopMapReduceCommitProtocol` set up a committer of `FileOutputCommitter` to deal with actual commit task / job. In `InsertIntoHadoopFsRelationCommand`, we pass `OutputSpec` to `FileFormatWriter.write`, this will trigger `FileOutputFormat.setOutputPath(job, new Path(outputSpec.outputPath))`, which would set `outputPath` of `FileOutputCommitter`, this is a `intermediatePath` rather than a `finalPath`.

Back to `SQLHadoopMapReduceCommitProtocol`, when `spark.sql.sources.outputCommitterClass` is set in conf and the specified committer class is assignable from `FileOutputCommitter`, the `outputPath` of this committer should be `intermediatePath`, but `outputPath` of `SQLHadoopMapReduceCommitProtocol` should be `finalPath`

Reply:
I see. I got your point.

Reply:
Hey @WinkerDu - thank you for the PR. One question: when dynamicPartitionOverwrite is on, this code block will only execute when `clazz` is non-null, which means SQLConf.OUTPUT_COMMITTER_CLASS is set. It works for parquet files since that SQL property is set at https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala#L97. So what about other file formats, like Orc? There seems to be no such property set logic for other file formats, at lease in Spark repo. So is dynamicPartitionOverwrite supposed to be for Parquet only? Am I missing sth here? Thanks. 

@cloud-fan @Ngone51 @agrawaldevesh